examples/text_style_transfer/main.py

```diff
@@ -0,0 +1,210 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Text style transfer
+
+This is a simplified implementation of:
+
+Toward Controlled Generation of Text, ICML2017
+Zhiting Hu, Zichao Yang, Xiaodan Liang, Ruslan Salakhutdinov, Eric Xing
+
+Download the data with the cmd:
+
+$ python prepare_data.py
+
+Train the model with the cmd:
+
+$ python main.py --config config
+"""
+
+# pylint: disable=invalid-name, too-many-locals, too-many-arguments, no-member
+
+import os
+import importlib
+import argparse
+import numpy as np
+import torch
+
+import texar.torch as tx
+
+from ctrl_gen_model import CtrlGenModel
+
+parser = argparse.ArgumentParser()
+
+parser.add_argument('--config', default='config', help="The config to use.")
+
+args = parser.parse_args()
+
+config = importlib.import_module(args.config)
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+def _main():
+    # Data
+    train_data = tx.data.MultiAlignedData(hparams=config.train_data,
+                                          device=device)
+    val_data = tx.data.MultiAlignedData(hparams=config.val_data,
+                                        device=device)
+    test_data = tx.data.MultiAlignedData(hparams=config.test_data,
+                                         device=device)
+    vocab = train_data.vocab(0)
+
+    # Each training batch is used twice: once for updating the generator and
+    # once for updating the discriminator. Feedable data iterator is used for
+    # such case.
+    iterator = tx.data.DataIterator(
+        {'train_g': train_data, 'train_d': train_data,
+         'val': val_data, 'test': test_data})
+
+    # Model
+    gamma_ = 1.
+    lambda_g_ = 0.
+
+    # Model
+    model = CtrlGenModel(vocab, hparams=config.model)
+    model.to(device)
+
+    # create optimizers
+    train_op_d = tx.core.get_optimizer(
+        params=model.d_vars,
+        hparams=config.model['opt']
+    )
+
+    train_op_g = tx.core.get_optimizer(
+        params=model.g_vars,
+        hparams=config.model['opt']
+    )
+
+    train_op_g_ae = tx.core.get_optimizer(
+        params=model.g_vars,
+        hparams=config.model['opt']
+    )
+
+    def _train_epoch(gamma_, lambda_g_, epoch, verbose=True):
+        model.train()
+        avg_meters_d = tx.utils.AverageRecorder(size=10)
+        avg_meters_g = tx.utils.AverageRecorder(size=10)
+        iterator.switch_to_dataset("train_g")
+        step = 0
+        for batch in iterator:
+            train_op_d.zero_grad()
+            train_op_g_ae.zero_grad()
+            train_op_g.zero_grad()
+            step += 1
+
+            vals_d = model(batch, gamma_, lambda_g_, mode="train",
```

Comment:
Which dataset do you use here? `train_g`? It seems that `train_d` is used. Why does such a difference exist?

Reply:
For training, I use the same dataset because it is tricky to switch datasets between steps. And considering the discriminator and generator are trained separately, I've noticed it doesn't really affect the results.

Reply:
Removing the 2 iterators and keeping just one train iterator.

Reply:
If that is the case, we do not need to have `train_d` in the iterator (?)